packages/signature-controller/src/SignatureController.ts

```diff
@@ -0,0 +1,680 @@
+    }
+
+    messageManager.rejectMessage(messageId);
+    this._rejectApproval(messageId);
+
+    return this._getState();
+  }
+
+  private _handleMessageManagerEvents<
+    M extends AbstractMessage,
+    P extends AbstractMessageParams,
+    PM extends AbstractMessageParamsMetamask,
+  >(
+    messageManager: AbstractMessageManager<M, P, PM>,
+    methodName: string,
+    eventName: string,
+  ) {
+    messageManager.hub.on('updateBadge', () => {
+      this.hub.emit('updateBadge');
+    });
+
+    messageManager.hub.on(
+      'unapprovedMessage',
+      (msgParams: AbstractMessageParamsMetamask) => {
+        this.hub.emit(eventName, msgParams);
+        this._requestApproval(msgParams, methodName);
+      },
+    );
+  }
+
+  private _subscribeToMessageState<
+    M extends AbstractMessage,
+    P extends AbstractMessageParams,
+    PM extends AbstractMessageParamsMetamask,
+  >(
+    messageManager: AbstractMessageManager<M, P, PM>,
+    updateState: (
+      state: SignatureControllerState,
+      newMessages: Record<string, StateMessage>,
+      messageCount: number,
+    ) => void,
+  ) {
+    messageManager.subscribe((state: MessageManagerState<AbstractMessage>) => {
+      const newMessages = this._migrateMessages(
+        state.unapprovedMessages as any,
+      );
+
+      this.update((draftState) => {
+        updateState(draftState, newMessages, state.unapprovedMessagesCount);
+      });
+    });
+  }
+
+  private _migrateMessages(
+    coreMessages: Record<string, CoreMessage>,
+  ): Record<string, StateMessage> {
+    const stateMessages: Record<string, StateMessage> = {};
+
+    for (const messageId of Object.keys(coreMessages)) {
+      const coreMessage = coreMessages[messageId];
+      const stateMessage = this._migrateMessage(coreMessage);
+
+      stateMessages[messageId] = stateMessage;
+    }
+
+    return stateMessages;
+  }
+
+  private _migrateMessage(coreMessage: CoreMessage): StateMessage {
+    const { messageParams, ...coreMessageData } = coreMessage;
+
+    // Core message managers use messageParams but frontend uses msgParams with lots of references
+    const stateMessage = {
+      ...coreMessageData,
+      msgParams: messageParams,
+    };
+
+    return stateMessage as StateMessage;
+  }
+
+  private _normalizeMsgData(data: string) {
+    if (data.slice(0, 2) === '0x') {
+      // data is already hex
+      return data;
+    }
+    // data is unicode, convert to hex
+    return bufferToHex(Buffer.from(data, 'utf8'));
+  }
+
+  private _getMessage(messageId: string): StateMessage {
+    return {
+      ...this.state.unapprovedMsgs,
+      ...this.state.unapprovedPersonalMsgs,
+      ...this.state.unapprovedTypedMessages,
+    }[messageId];
+  }
+
+  private _requestApproval(
+    msgParams: AbstractMessageParamsMetamask,
+    type: string,
+  ) {
+    const id = msgParams.metamaskId as string;
+    const origin = msgParams.origin ?? ORIGIN_METAMASK;
```

Comment:
I'm unsure about what exact behaviour we want here, but with `??`, `ORIGIN_METAMASK` will be used only if `msgParams.origin` is `undefined`.

If we were to use `||` it'll be used if `msgParams.origin` is falsy, such as `''`.

Reply:
I was trying to be more explicit to show we're handling an undefined origin, but you're right, we should handle the empty string case too. Done.